### PR TITLE
fix(saml): validate SSO tokens against the schema that mints them

### PR DESCRIPTION
**Pull request by agrim123**
## **User description**
A completed SSO login was refused by the API. Observed on GCP staging:

```
saml login succeeded      user_id=user_01M09SRKR4PJS5ZY68KC2V9J2N
failed to validate token  error="token missing user ID"
  internal/auth.(*supabaseAuth).ValidateToken  supabase.go:125
```

The assertion validated, the token was minted, and then every request after it
came back 401.

## Cause

The mint and the verify used different claim schemas.

| | user claim | tenant claim |
|---|---|---|
| SAML minted (Flexprice provider) | `user_id` | `tenant_id` |
| Middleware verified (Supabase provider) | `sub` | `app_metadata.tenant_id` |

The SAML handler minted through the built-in Flexprice provider, while
`AuthenticateMiddleware` builds its validator from `auth.provider` — Supabase on
that deployment. Both sign with the same `auth.secret`, so the signature was
fine; the claim names were not.

Delegating to the configured provider instead is not an option: an assertion
authenticates a user that exists only in our database, so there is no Supabase
account for Supabase to have issued a token for. SSO needs its own issuer and
validator, which is what this adds.

## Security

This widens what the middleware accepts, so the reasoning is worth stating
plainly.

Both providers sign with the same secret, so **a signature alone cannot prove a
token came from the SSO flow**, and the marker claim travels inside the token
where anyone able to mint one could set it. The marker is therefore never
treated as authority. It selects which validator runs; that validator then
independently requires all of:

1. `auth.saml.enabled` on the deployment — config, not caller-controlled
2. a valid signature, algorithm **pinned to HMAC** (blocks `alg=none` and
   key-confusion forgeries)
3. the marker, re-checked after verification
4. non-empty `user_id` and `tenant_id`

**Routing to the SSO validator can never be an advantage**, because it is
strictly stricter than the ordinary one. A caller who bolts the marker onto a
token the ordinary provider would have accepted only gets their own token
refused.

Two existing safeguards still apply on top: the middleware loads the named user
and refuses one archived or missing, and the session's roles come from that user
record rather than from any claim. A token can name an identity but never grant
itself privileges.

**Blast radius:** a deployment with `auth.saml.enabled=false` accepts no SSO
token at all, so enabling SAML is the only thing that changes what the API
accepts. Existing password logins are untouched — a token without the marker
takes exactly the path it takes today (covered by a test).

## Tests

Round trip under **both** providers, plus adversarial cases:

- marker bolted onto a Supabase-shaped token the ordinary validator would accept
- a valid SSO token replayed at a deployment with SAML off
- a session-shaped token carrying the marker
- wrong secret across HS256 / HS384 / HS512
- `alg=none`
- expired token
- missing / empty identity claims
- marker type confusion (`"true"`, `1`)
- an ordinary token is still routed to the ordinary validator

## Verification

- `go build ./internal/...` clean
- `make lint-ci` (loglint merge gate) exit 0
- full `go test ./internal/...` — **61 ok, 0 FAIL**, exit 0
- rebased on `develop` and re-verified after (fresh `-count=1` run, not cached)

`gofmt -l` flags four files in the touched packages; all four are identical on
`develop`, so they are pre-existing and deliberately not reformatted here.

The SAML ACS handler was the only production caller of `GenerateDevToken`, so
nothing else changes behaviour.

## Known gap, not addressed here

`enforce_sso` has **no effect on a Supabase-backed deployment**.
`refusePasswordLoginUnderSSO` is reached only from `authService.Login`, but such
a deployment's frontend calls `supabase.auth.signInWithPassword()` directly and
never touches that path. A tenant enabling enforcement today still has every
user able to sign in with a password, bypassing their identity provider.

Fixing it means moving the check to `AuthenticateMiddleware` — the one chokepoint
both login paths cross — which changes behaviour for every authenticated request
and belongs in its own PR. Until then `enforce_sso` should be treated as
additive-only on hosted deployments, not offered as an enforcement control.

Related: `resolveUser` logs in any existing user with a matching email without
regard to how that account was created, so one person can hold both a password
and an IdP identity with no way to retire the password.

___

## **CodeAnt-AI Description**
Restore authenticated requests after SAML login across supported auth providers

### What Changed
- SAML login now issues tokens with the identity claims expected by SSO validation, so users can access the API after signing in.
- SSO tokens are validated separately from password-login tokens, allowing SAML to work when Supabase handles regular password authentication.
- SSO tokens are rejected when SAML is disabled, expired, unsigned, incorrectly signed, missing identity data, or missing the SSO marker.
- Added coverage for valid SSO login flows, ordinary token handling, and forged or malformed tokens.

### Impact
`✅ SAML users can make API requests after login`
`✅ SSO works alongside Supabase password authentication`
`✅ Fewer unauthorized responses after successful SSO login`
<details><summary><strong>💡 Usage Guide</strong></summary>

### Checking Your Pull Request
Every time you make a pull request, our system automatically looks through it. We check for security issues, mistakes in how you're setting up your infrastructure, and common code problems. We do this to make sure your changes are solid and won't cause any trouble later.

### Talking to CodeAnt AI
Got a question or need a hand with something in your pull request? You can easily get in touch with CodeAnt AI right here. Just type the following in a comment on your pull request, and replace "Your question here" with whatever you want to ask:
<pre>
<code>@codeant-ai ask: Your question here</code>
</pre>
This lets you have a chat with CodeAnt AI about your pull request, making it easier to understand and improve your code.

#### Example
<pre>
<code>@codeant-ai ask: Can you suggest a safer alternative to storing this secret?</code>
</pre>

### Preserve Org Learnings with CodeAnt
You can record team preferences so CodeAnt AI applies them in future reviews. Reply directly to the specific CodeAnt AI suggestion (in the same thread) and replace "Your feedback here" with your input:
<pre>
<code>@codeant-ai: Your feedback here</code>
</pre>
This helps CodeAnt AI learn and adapt to your team's coding style and standards.

#### Example
<pre>
<code>@codeant-ai: Do not flag unused imports.</code>
</pre>

### Retrigger review
Ask CodeAnt AI to review the PR again, by typing:
<pre>
<code>@codeant-ai: review</code>
</pre>

### Check Your Repository Health
To analyze the health of your code repository, visit our dashboard at [https://app.codeant.ai](https://app.codeant.ai). This tool helps you identify potential issues and areas for improvement in your codebase, ensuring your repository maintains high standards of code health.

</details>

## Summary by CodeRabbit

* **New Features**
  * Added secure SSO token issuance and validation for SAML authentication.
  * SSO tokens include tenant and user identities, expiration, and signature protection.
  * Authentication automatically routes SSO tokens through the appropriate validation process.

* **Bug Fixes**
  * Improved rejection of expired, unsigned, incorrectly signed, malformed, or misclassified tokens.
  * Prevented SSO tokens from being accepted when SAML authentication is disabled.
  * SAML authentication now requires a configured signing secret at startup.

* **Tests**
  * Added comprehensive coverage for valid flows, token separation, identity requirements, configuration, and security edge cases.

### CODE REVIEW

File: internal/auth/sso_token.go [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+			Mark(ierr.ErrPermissionDenied)
+	}
+
+	tenantID, _ := claims[ssoClaimTenantID].(string)
+	userID, _ := claims[ssoClaimUserID].(string)
+	if strings.TrimSpace(tenantID) == "" || strings.TrimSpace(userID) == "" {
+		return nil, ierr.NewError("token missing identity claims").
+			WithHint("Invalid token").
+			Mark(ierr.ErrPermissionDenied)
+	}
```
[the remark this answers is not in the document]
Comment (agrim123) [reply internal/auth/sso_token.go:172]: Confirmed and fixed in 42efbb424 — this was real and exploitable.

I verified it before fixing rather than taking the finding at face value. With `auth.secret` empty, a token signed with `[]byte("")` authenticated as an arbitrary user in an arbitrary tenant:

```
FORGERY SUCCEEDED: authenticated as victim_tenant/victim_superadmin with an empty secret
```

Your reading of why nothing caught it is exactly right: `validateSecrets` is warn-only (deliberately, after the 2026-07-06 full-struct-validation crashloop), and it checks `auth.secret` only under the Flexprice provider — so a Supabase deployment with SSO enabled and no secret booted silently.

Fixed in `validateSAMLDependencies` beside the Redis requirement rather than in `validateSecrets`. That placement scopes the hard failure to SSO being enabled, so it cannot take down a deployment that does not offer SSO — the same reasoning that keeps the general secret validation warn-only.

```go
if strings.TrimSpace(c.Auth.Secret) == "" {
    return fmt.Errorf("auth.saml.enabled requires a non-empty auth.secret (FLEXPRICE_AUTH_SECRET): " +
        "it signs the SSO token, and an empty key lets anyone mint one naming any user")
}
```

Covers `""`, `"   "` and `"\t"`, since a whitespace-only key is no better than an empty one.

Worth flagging for reviewers: this also surfaced an existing `TestValidateSAMLDependencies` that was exercising SAML dependencies with no secret set. I updated the test to set one rather than relaxing the check to keep it passing.

File: internal/auth/sso_token.go [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+	expiresAt := time.Now().Add(time.Duration(expiryHours) * time.Hour)
+	token := jwt.NewWithClaims(jwt.SigningMethodHS256, jwt.MapClaims{
+		ssoClaimTenantID: tenantID,
+		ssoClaimUserID:   userID,
+		ssoTokenClaim:    true,
+		"exp":            expiresAt.Unix(),
+		"iat":            time.Now().Unix(),
+	})
```
[the remark this answers is not in the document]
Comment (agrim123) [reply internal/auth/sso_token.go:81]: Fixed in 42efbb424, including your `now` suggestion.

Not reachable today — the only caller passes the `tokenExpiryHours` constant (`24 * 30`) — but the guard is cheap and the failure mode you describe is precisely the one this type exists to prevent: the mint succeeds, the caller redirects with a future `expiresAt`, and the browser gets a 401 on the first dashboard request with nothing pointing at the expiry. That is the same mint-succeeds-verify-fails shape as the bug this PR fixes, and it would read as an identity-provider fault rather than a configuration one.

Also took the single-clock-reading change: with a small enough expiry, reading `time.Now()` twice can put `iat` after `exp`.

Added `TestSSOIssuerRejectsNonPositiveExpiry` covering `0`, `-1` and `-24`.

**Review by DiabolusGX — APPROVED** [review]
